File: data-structure/disjoint-set/disjointSet.js
class DisjointSet {
    constructor(n) {
        this.p = new Array(n).fill(0);
        this.init(n);
    }

    init(n) {
        for (let i = 1; i <= n; i++) {
            this.makeSet(i);
        }
    }

    makeSet(x) {
        this.p[x] = x;
    }

    findSet(x) {
        if (this.p[x] == x) {
            return this.p[x]
        }

        this.p[x] = this.findSet(this.p[x])
        return this.p[x]
    }

    unionSet(x, y) {
        const px = this.findSet(x);
        const py = this.findSet(y);

        this.p[py] = px;
    }
}

const disjointSet = new DisjointSet(8);
disjointSet.unionSet(1, 3);
disjointSet.unionSet(2, 3);
disjointSet.unionSet(5, 6);
disjointSet.unionSet(6, 8);
disjointSet.unionSet(1, 5);
disjointSet.unionSet(6, 7);
console.log(disjointSet);

File: data-structure/disjoint-set/DisjointSet.java
package com.example;

public class DisjointSet {
    private final int[] path;
    private final int[] height;

    public DisjointSet(int size) {
        path = new int[size];
        height = new int[size];

        for (int i = 1; i < size; i++) {
            path[i] = i;
        }
    }

    public int findSet(int a) {
        if (path[a] != a) {
            path[a] = findSet(path[a]);
        }
        return path[a];
    }

    public void unionSet(int a, int b) {
        int px = findSet(a);
        int py = findSet(b);

        if (height[px] >= height[py]) {
            path[py] = px;
            if (height[px] == height[py]) {
                height[px] += 1;
            }
        } else {
            path[px] = py;
        }
    }
}
